### PR TITLE
CQLin for Conv and MatMul

### PR DESCRIPTION
**Pull request by liliatangxy**
Use CQLin when possible for Conv and MatMul

square matrices:
matmul 32x32: 61.685146ms
cqlin 32x32: 32.491135

matmul 1024x1024: 200.716086ms
cqlin 1024x1024: 133.131949ms

### CODE REVIEW

**Comment by ddkang** [comment]
Do we have any benchmarks on this?

**Comment by liliatangxy** [comment]
I put up the ones I collected before the meeting. I can do comparisons for gptj and cnn layers once I fix the test bug

File: src/basic_block/cqlin.rs [hunk 1]
```diff
@@ -6,19 +6,35 @@ use ark_bn254::{Fr, G1Affine, G1Projective, G2Affine, G2Projective};
...
 impl BasicBlock for CQLinBasicBlock {
+  fn genModel(&self) -> ArrayD<Fr> {
+    self.setup.clone()
+  }
+
   fn run(&self, model: &ArrayD<Fr>, inputs: &Vec<&ArrayD<Fr>>) -> Result<Vec<ArrayD<Fr>>, util::CQOutOfRangeError> {
-    assert!(model.ndim() == 2 && inputs.len() == 1 && inputs[0].ndim() == 2 && inputs[0].shape()[1] == model.shape()[0]);
+    assert!(model.ndim() == 2);
+    assert!(inputs.len() == 1);
+    assert!(
```
Comment (Vimalakirti) [inline src/basic_block/cqlin.rs:28]: What is the reason to add this assertion?

File: src/layer/matmul.rs [hunk 2]
```diff
@@ -13,24 +13,17 @@ impl Layer for MatMulLayer {
   fn graph(
     input_shapes: &Vec<&Vec<usize>>,
     input_types: &Vec<DatumType>,
-    _constants: &Vec<Option<(&ArrayD<Fr>, DatumType)>>,
+    constants: &Vec<Option<(&ArrayD<Fr>, DatumType)>>,
     _attributes: &Vec<&AttributeProto>,
   ) -> (Graph, Vec<Vec<usize>>, Vec<DatumType>) {
     let mut graph = Graph::new();
+    println!("input_shapes {:?}", input_shapes);
```
Comment (Vimalakirti) [inline src/layer/matmul.rs]: please remove this `println!`

File: src/basic_block/cqlin.rs (hunk 1 above)
Comment (liliatangxy) [reply src/basic_block/cqlin.rs:28]: In the case that the input is 1d, we still want to enable using cqlin by treating it as a 2d 1xn array

Comment (Vimalakirti) [reply src/basic_block/cqlin.rs:28]: I see, does this affect the output shape? (do we need to squeeze the additional dim 1 after we complete the cqlin?)

Comment (liliatangxy) [reply src/basic_block/cqlin.rs:28]: fixed

Comment (Vimalakirti) [reply src/basic_block/cqlin.rs:28]: thanks

File: src/tests.rs [hunk 3]
```diff
@@ -108,7 +108,7 @@ fn testBasicBlocks() {
   let b = ArrayD::from_shape_fn(IxDyn(&[n, m]), |_| Fr::rand(&mut rng));
   let c = ArrayD::from_shape_fn(IxDyn(&[m, n]), |_| Fr::rand(&mut rng));
   testBasicBlock(MatMulBasicBlock {}, srs, &empty, &vec![&a, &b]);
-  testBasicBlock(CQLinBasicBlock {}, srs, &c, &vec![&a]);
+  testBasicBlock(CQLinBasicBlock { setup: c.clone() }, srs, &c, &vec![&a]);
```
Comment (Vimalakirti) [inline src/tests.rs:113]: Do you think it is good to test both 2D and 1D inputs here?

**Review by Vimalakirti — DISMISSED** [review]

File: src/basic_block/cqlin.rs [hunk 4]
```diff
@@ -6,21 +6,42 @@ use ark_bn254::{Fr, G1Affine, G1Projective, G2Affine, G2Projective};
...
```
Comment (ddkang) [inline src/basic_block/cqlin.rs:28]: This needs a comment

**Review by Vimalakirti — APPROVED** [review]

**Comment by ddkang** [comment]
Are there any regressions with the logic changes?

**Comment by liliatangxy** [comment]
I tested on layers with 1d, 2d, and 3d inputs for correctness.

**Comment by ddkang** [comment]
What about performance?

**Comment by liliatangxy** [comment]
GPTj Matmul layer was ~10x faster and using CQLin in Conv layer prevents us from needing to prove the weight splatting with copy constraint.

**Comment by ddkang** [comment]
I meant does it affect the performance of non-cqlin matrix multiplication

**Comment by liliatangxy** [comment]
No

**Review by ddkang — APPROVED** [review]

**Pull request merged**